### PR TITLE
Homework 7

### PR DESCRIPTION
**Pull request by kdr1mkf**
Homework 7 seems to be done.

### CODE REVIEW

File: src/test/java/com/epam/tc/hw7/components/HeaderMenu.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.epam.tc.hw7.components;
+
+import com.epam.jdi.light.elements.complex.Menu;
+import com.epam.jdi.light.elements.pageobjects.annotations.locators.UI;
+import com.epam.jdi.light.ui.html.elements.common.Link;
+
+public class HeaderMenu extends Menu {
+
+    @UI("//*[contains(@class, \"m-l8\")]//*[contains(text(), \"Metals & Colors\")]")
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/components/HeaderMenu.java:9]: Locator is very difficult. 
The correct way is to write locator for header and simple locator for page by text

File: src/test/java/com/epam/tc/hw7/components/LogInForm.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    @UI("#user-name")
+    protected static Text fullName;
+    @UI("#name")
+    protected static TextField name;
+    @UI("#password")
+    protected static TextField password;
+    @UI("#login-button")
+    protected static Button loginButton;
+
+    public void logIn(User user) {
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/components/LogInForm.java:23]: This function is not for Form. Fullname is not a part of the form, so we break the logic.

File: src/test/java/com/epam/tc/hw7/components/LogInForm.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        if (!fullName.isDisplayed()) {
+            userIcon.click();
+            name.setValue(user.getName());
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/components/LogInForm.java:26]: There is a function loginAs in Form, it's a part of that function that we can just use

File: src/test/java/com/epam/tc/hw7/components/LogInForm.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import com.epam.jdi.light.elements.pageobjects.annotations.locators.UI;
+import com.epam.jdi.light.ui.html.elements.common.Button;
+import com.epam.jdi.light.ui.html.elements.common.Text;
+import com.epam.jdi.light.ui.html.elements.common.TextField;
+import com.epam.tc.hw7.domain.User;
+
+public class LogInForm extends Form<User> {
+
+    @UI("#user-icon")
+    protected static Button userIcon;
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/components/LogInForm.java:13]: This is not a part of Form

File: src/test/java/com/epam/tc/hw7/domain/DataSet.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package com.epam.tc.hw7.domain;
+
+import com.epam.jdi.tools.DataClass;
+import java.util.List;
+
+public class DataSet extends DataClass<DataSet> {
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/domain/DataSet.java:6]: DataSet is a very common name, it's better to rename it

File: src/test/java/com/epam/tc/hw7/page/JdiLightSite.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    @Url("/index.html") @Title("Home Page")
+    protected static HomePage homePage;
+    @Url("/metals-colors.html") @Title("Metal and Colors")
+    protected static MetalsAndColorsPage metalsAndColorsPage;
+
+    public static void logIn(User user) {
+        homePage.logIn(user);
+    }
+
+    public static void fillMetalsAndColorsForm(DataSet dataSet) {
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/page/JdiLightSite.java:22]: This function cann't be a part of Site. At this place we can only create a functions that can be used in ANY page

File: src/test/java/com/epam/tc/hw7/page/JdiLightSite.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        metalsAndColorsPage.fillMetalsAndColorsForm(dataSet);
+    }
+
+    public static List<String> getResultMessages() {
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/page/JdiLightSite.java:26]: This function cann't be a part of Site. At this place we can only create a functions that can be used in ANY page

File: src/test/java/com/epam/tc/hw7/page/JdiLightSite.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    public static void navigateToMetalsAndColorsPage() {
+        homePage.navigateToMetalsAndColorsPage();
+        metalsAndColorsPage.checkOpened();
+    }
+
+    public static void openHomePage() {
+        homePage.openPage();
+    }
+
+    public static void submitMetalsAndColorsForm() {
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/page/JdiLightSite.java:39]: This function cann't be a part of Site. At this place we can only create a functions that can be used in ANY page

File: src/test/java/com/epam/tc/hw7/test/MetalsAndColorsFormTest.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.List;
+import org.assertj.core.api.Assertions;
+import org.testng.annotations.Test;
+
+public class MetalsAndColorsFormTest extends CommonTest {
+
+    @Test(description = "Metals & colors form test",
+          dataProvider = "Valid data", dataProviderClass = MetalsAndColorsDataProvider.class)
+    public void testWithValidData(User user, DataSet dataSet) {
+        JdiLightSite.logIn(user);
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/test/MetalsAndColorsFormTest.java:17]: Login is not a part of the test, it's a precondition

**Review by pnatashap — CHANGES_REQUESTED** [review]
Comments are in a code

File: src/test/java/com/epam/tc/hw7/test/MetalsAndColorsFormTest.java (hunk 9 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/test/MetalsAndColorsFormTest.java:17]: LogIn is not a precondition. Log in is marked as a step of the test  in the task. My approach meets the task.
login-as-step

File: src/test/java/com/epam/tc/hw7/components/HeaderMenu.java (hunk 1 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/components/HeaderMenu.java:9]: I see no any kind of "difficulty" in the locator. Anyway, it's fixed.

File: src/test/java/com/epam/tc/hw7/components/LogInForm.java (hunk 4 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/components/LogInForm.java:13]: Fixed.

File: src/test/java/com/epam/tc/hw7/domain/DataSet.java (hunk 5 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/domain/DataSet.java:6]: Fixed.

File: src/test/java/com/epam/tc/hw7/components/LogInForm.java (hunk 3 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/components/LogInForm.java:26]: The comment above says 
>  this function is not for Form

This comment says
>  There is a function loginAs in Form

Don't these comments break the logic?

loginAs brings me another problem. loginAs tryes to find and fill "fullName" field. I have no idea where I'm supposed to store the fullName attribute that belongs to a User entitiy. If I'm supposed to override loginAs logic anyway, why do I need loginAs then?

File: src/test/java/com/epam/tc/hw7/page/JdiLightSite.java (hunk 6 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/page/JdiLightSite.java:22]: We can use that function at any place of the site. It throws an exception if the method is used in the wrong way. How am I supposed to interract with pages? Direct access to the class attributes violates incapsulation.
Direct access provides the opportunity to use any method at any page and does not solve the mentioned problem.

File: src/test/java/com/epam/tc/hw7/page/JdiLightSite.java (hunk 7 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/page/JdiLightSite.java:26]: We can use that function at any place of the site. It throws an exception if the method is used in the wrong way. How am I supposed to interract with pages? Direct access to the class attributes violates incapsulation.
Direct access provides the opportunity to use any method at any page and does not solve the mentioned problem.

File: src/test/java/com/epam/tc/hw7/page/JdiLightSite.java (hunk 8 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/page/JdiLightSite.java:39]: We can use that function at any place of the site. It throws an exception if the method is used in the wrong way. How am I supposed to interract with pages? Direct access to the class attributes violates incapsulation.
Direct access provides the opportunity to use any method at any page and does not solve the mentioned problem.

File: src/test/java/com/epam/tc/hw7/components/LogInForm.java (hunk 2 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/components/LogInForm.java:23]: Fixed.

File: src/test/java/com/epam/tc/hw7/components/LogInForm.java (hunk 3 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/components/LogInForm.java:26]: Anyway fixed.

File: src/test/java/com/epam/tc/hw7/page/JdiLightSite.java (hunk 8 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/page/JdiLightSite.java:39]: Fixed.

File: src/test/java/com/epam/tc/hw7/page/JdiLightSite.java (hunk 7 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/page/JdiLightSite.java:26]: Fixed.

File: src/test/java/com/epam/tc/hw7/page/JdiLightSite.java (hunk 6 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/page/JdiLightSite.java:22]: Fixed.

File: src/test/java/com/epam/tc/hw7/page/HomePage.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    public void logIn(User user) {
+        if (!fullName.isDisplayed()) {
+            userIcon.click();
+            loginForm.loginAs(user);
+        }
+        assertLoggedIn(user);
+    }
+
+    protected void assertLoggedIn(User user) {
+        fullName.assertThat().displayed();
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/page/HomePage.java:31]: Function that doesn't use a parameter - bad designed function.
There is no actual assert that this user is loggedIn

File: src/test/java/com/epam/tc/hw7/test/CommonTest.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    @BeforeSuite(alwaysRun = true)
+    static void setUpSuite() {
+        initSite(JdiLightSite.class);
+    }
+
+    @BeforeMethod(alwaysRun = true)
+    public void setUp() {
+        JdiLightSite.openHomePage();
+        User user = new UserReader()
+            .readJsonToListOfObjects(".\\src\\test\\resources\\hw7\\users.json")
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/test/CommonTest.java:24]: Bad practice to read file for EVERY test, operation should be performed only once

File: src/test/java/com/epam/tc/hw7/test/CommonTest.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+            .get(0);
+        JdiLightSite.logIn(user);
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/test/CommonTest.java:26]: If you read a test, you can see that it's not correct to logIn every test. It's better to make sure that we are logged in.

**Review by pnatashap — CHANGES_REQUESTED** [review]
Comments are in code

File: src/test/java/com/epam/tc/hw7/page/HomePage.java (hunk 10 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/page/HomePage.java:31]: This is the consequence of fixes that were required before.

I can not store user's fullname (JDI fill form feature) in the User class (see requred changes before), so I have no user's fullname. So I can not check if a particular user is logged in.

File: src/test/java/com/epam/tc/hw7/test/CommonTest.java (hunk 11 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/test/CommonTest.java:24]: What if I want to run the test for different users?

File: src/test/java/com/epam/tc/hw7/test/CommonTest.java (hunk 12 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/test/CommonTest.java:26]: Login that's been performed may influence the next test. I have no information what the "better" is like. I have a test case, that's supposed to be automated and log in is a step.

BTW I do not log in every time. JdiLightSite invokes HomePage method to log in. The HomePage method checks if a user is logged in.

I have no idea what am I supposed to do. Do I have to provide a direct access to the class attributes? Is this a best practice?

File: src/test/java/com/epam/tc/hw7/test/CommonTest.java (hunk 11 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw7/test/CommonTest.java:24]: Whatever.

**Review by pnatashap — APPROVED** [review]

**Pull request merged**